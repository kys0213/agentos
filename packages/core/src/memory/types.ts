export type NodeType = 'query'|'answer'|'feedback'|'entity';
export type EdgeType = 'similar_to'|'responded_with'|'has_feedback'|'refers_to_entity';

export interface BaseNode {
  id: string;
  type: NodeType;
  text?: string;
  canonicalKey?: string;
  embedding?: number[] | Map<number, number>;
  createdAt: number;
  lastAccess: number;
  weights: { repeat: number; feedback: number };
  degree: number;
  pinned?: boolean;
}

export interface Edge {
  id: string;
  from: string;
  to: string;
  type: EdgeType;
  weight: number;
  createdAt: number;
  lastAccess: number;
}

export interface GraphConfig {
  maxNodes: number;
  maxEdges: number;
  halfLifeMin: number;
  tauDup: number;
  tauSim: number;
  protectMinDegree: number;
  enableInvertedIndex?: boolean;
}

export interface EmbeddingProvider {
  embed(text: string): number[] | Map<number, number>;
  exportState?(): any;
  importState?(s: any): void;
}

