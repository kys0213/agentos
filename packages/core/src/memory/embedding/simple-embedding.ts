import type { EmbeddingProvider } from '../../memory/types';

export type SparseVec = Map<number, number>;

const NF = (s: string) => s.normalize('NFC');
export function normalize(text: string) {
  return NF(text.toLowerCase())
    .replace(/[^\p{L}\p{N}\s.@/_-]/gu, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

function ngrams(s: string, minN = 3, maxN = 5): string[] {
  const out: string[] = [];
  const str = s.replace(/\s+/g, ' ');
  for (let n=minN; n<=maxN; n++) {
    for (let i=0; i<=str.length-n; i++) out.push(str.slice(i, i+n));
  }
  return out;
}

function hash64(s: string, seed = 0xC0FFEE): number {
  let h = BigInt(seed) ^ 0xcbf29ce484222325n;
  for (let i=0; i<s.length; i++) {
    h ^= BigInt(s.charCodeAt(i));
    h = (h * 0x00000100000001B3n) & 0xFFFFFFFFFFFFFFFFn;
  }
  return Number(h & 0xFFFFFFFFn);
}

export class SimpleEmbedding implements EmbeddingProvider {
  constructor(private dim = 16384, private minN = 3, private maxN = 5, private seed = 0xC0FFEE) {}

  exportState() {
    return { kind: 'ngram-hash', dim: this.dim, minN: this.minN, maxN: this.maxN, seed: this.seed, normVersion: 'nfc_v1', hashAlgo: 'fnv1a64_demo' };
  }
  importState(s: any) {
    if (!s) return;
    this.dim = s.dim ?? this.dim; this.minN = s.minN ?? this.minN; this.maxN = s.maxN ?? this.maxN; this.seed = s.seed ?? this.seed;
  }

  embed(text: string): SparseVec {
    const t = normalize(text);
    const vec = new Map<number, number>();
    const grams = ngrams(t, this.minN, this.maxN);
    for (const g of grams) {
      const idx = Math.abs(hash64(g, this.seed)) % this.dim;
      vec.set(idx, (vec.get(idx) ?? 0) + 1);
    }
    let norm = 0; for (const v of vec.values()) norm += v*v; norm = Math.sqrt(norm) || 1;
    for (const [i,v] of vec) vec.set(i, v/norm);
    return vec;
  }
}

