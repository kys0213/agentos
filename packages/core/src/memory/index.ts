export * from './graph-store';
export * from './memory-orchestrator';
export * from './types';
