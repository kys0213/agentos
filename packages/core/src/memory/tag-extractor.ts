import { LlmBridge, UserMessage } from 'llm-bridge-spec';
import { parseJsonArray } from '@agentos/lang/json';

import { canonicalTagKey } from './utils/tag-key';

export interface TagExtractorInput {
  texts: string[];
  existing: string[];
  maxTags: number;
}

export interface TagExtractor {
  extract(input: TagExtractorInput): Promise<string[]>;
}

export interface LlmTagExtractorOptions {
  /**
   * 시스템 프롬프트는 호출 시마다 재사용되며, 한국어 대화 맥락에서 태그를 도출하도록 안내합니다.
   */
  systemPrompt?: string;
  /**
   * 대화 묶음과 기존 태그를 나열할 때 사용할 구분자. 기본값은 줄바꿈입니다.
   */
  joinSeparator?: string;
  /**
   * 결과 최대 태그 수의 기본값. 호출 입력의 maxTags가 우선합니다.
   */
  maxTags?: number;
}

const DEFAULT_SYSTEM_PROMPT =
  '당신은 대화 히스토리에서 의미 있는 주제 태그를 도출하는 한국어 어시스턴트입니다. ' +
  '사용자가 제공한 기존 태그 목록과 중복되지 않도록 주의하며, 출력은 반드시 JSON 배열 문자열 형식으로만 반환하세요.';

export class LlmTagExtractor implements TagExtractor {
  private readonly systemPrompt: string;
  private readonly separator: string;
  private readonly defaultMax: number;

  constructor(private readonly llm: LlmBridge, options?: LlmTagExtractorOptions) {
    this.systemPrompt = options?.systemPrompt ?? DEFAULT_SYSTEM_PROMPT;
    this.separator = options?.joinSeparator ?? '\n';
    this.defaultMax = options?.maxTags ?? 8;
  }

  async extract({ texts, existing, maxTags }: TagExtractorInput): Promise<string[]> {
    if (!texts.length) {
      return [];
    }

    const limit = Math.max(1, Math.min(maxTags || this.defaultMax, 64));

    const existingSet = new Set(existing.map((tag) => canonicalTagKey(tag)));

    const conversationBlock = texts
      .map((text, index) => `${index + 1}. ${text}`)
      .join(this.separator);
    const existingBlock = existing.length
      ? existing.map((tag) => `- ${tag}`).join(this.separator)
      : '- (없음)';

    const payload =
      '### 기존 태그 목록\n' +
      `${existingBlock}\n\n` +
      '### 최근 대화 묶음\n' +
      `${conversationBlock}\n\n` +
      `위 맥락을 대표하는 새로운 태그를 최대 ${limit}개까지 추출하세요. ` +
      '기존 태그와 의미가 매우 유사하면 중복 생성하지 마세요.';

    const messages: UserMessage[] = [
      {
        role: 'system',
        content: [{ contentType: 'text', value: this.systemPrompt }],
      },
      {
        role: 'user',
        content: [{ contentType: 'text', value: payload }],
      },
    ];

    const response = await this.llm.invoke({ messages });
    const raw = extractText(response);
    const parsed = parseJsonArray(raw);

    const seen = new Set(existingSet);
    const results: string[] = [];

    for (const kw of parsed) {
      if (results.length >= limit) {
        break;
      }
      const normalized = String(kw).trim();
      if (!normalized) {
        continue;
      }
      const key = canonicalTagKey(normalized);
      if (seen.has(key)) {
        continue;
      }
      seen.add(key);
      results.push(normalized);
    }

    return results;
  }
}

function extractText(resp: { content: { contentType: string; value: unknown } }): string {
  if (resp.content.contentType !== 'text') {
    throw new Error(`Unsupported content type: ${resp.content.contentType}`);
  }
  return String(resp.content.value ?? '');
}

