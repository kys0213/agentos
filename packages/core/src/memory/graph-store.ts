import fs from 'fs/promises';
import { BaseNode, Edge, EdgeType, EmbeddingProvider, GraphConfig } from './types';
import { SimpleEmbedding, SparseVec, normalize } from './embedding/simple-embedding';
import { cosineSparse } from './utils/cosine-sparse';
import { canonicalKey, defaultCanonicalMeta } from './utils/canonical-key';
import { serializeSparse, deserializeSparse } from './utils/serialize-sparse';

export class GraphStore {
  private nodes = new Map<string, BaseNode>();
  private edges = new Map<string, Edge>();
  private byCanonical = new Map<string, string>();
  private inverted?: Map<string, Set<string>>;
  private dirty = false;

  constructor(
    private cfg: GraphConfig,
    private embedder: EmbeddingProvider = new SimpleEmbedding()
  ) {
    if (cfg.enableInvertedIndex) {
      this.inverted = new Map();
    }
  }

  private indexNode(id: string, text?: string, canonical?: string) {
    if (canonical) {
      this.byCanonical.set(canonical, id);
    }
    if (!this.inverted || !text) {
      return;
    }
    const grams = normalize(text).split(' ');
    for (const g of grams) {
      if (!g) {
        continue;
      }
      const set = this.inverted.get(g) ?? new Set();
      set.add(id);
      this.inverted.set(g, set);
    }
  }

  upsertQuery(text: string) {
    const now = Date.now();
    const emb = this.embedder.embed(text) as SparseVec;
    const ck = canonicalKey(text);
    const existing = this.byCanonical.get(ck);
    if (existing) {
      const q = this.nodes.get(existing)!;
      q.weights.repeat += 1;
      q.lastAccess = now;
      this.touchNeighbors(existing, now);
      this.evictIfNeeded();
      return q.id;
    }
    const cand: { id: string; sim: number }[] = [];
    for (const n of this.nodes.values()) {
      if (n.type !== 'query' || !n.embedding) {
        continue;
      }
      const sim = cosineSparse(emb, n.embedding as SparseVec);
      cand.push({ id: n.id, sim });
    }
    cand.sort((a, b) => b.sim - a.sim);
    const best = cand[0];
    if (best && best.sim >= this.cfg.tauDup) {
      const q = this.nodes.get(best.id)!;
      q.weights.repeat += 1;
      q.lastAccess = now;
      this.touchNeighbors(q.id, now);
      this.evictIfNeeded();
      return q.id;
    }
    const id = `q_${now}_${Math.random().toString(36).slice(2, 8)}`;
    const q: BaseNode = {
      id,
      type: 'query',
      text,
      canonicalKey: ck,
      embedding: emb,
      createdAt: now,
      lastAccess: now,
      degree: 0,
      weights: { repeat: 0, feedback: 0 },
    };
    this.nodes.set(id, q);
    this.indexNode(id, text, ck);
    for (const c of cand.slice(0, 10)) {
      if (c.sim >= this.cfg.tauSim) {
        this.link(id, c.id, 'similar_to', c.sim);
        this.bump(id, +0.3);
        this.bump(c.id, +0.3);
      }
    }
    this.dirty = true;
    this.evictIfNeeded();
    return id;
  }

  recordFeedback(targetQueryId: string, label: 'up' | 'down' | 'retry', note?: string) {
    const now = Date.now();
    const id = `f_${now}_${Math.random().toString(36).slice(2, 6)}`;
    const f: BaseNode = {
      id,
      type: 'feedback',
      text: `feedback:${label}${note ? ' ' + note : ''}`,
      createdAt: now,
      lastAccess: now,
      degree: 0,
      weights: { repeat: 0, feedback: 0 },
    };
    this.nodes.set(id, f);
    this.link(targetQueryId, id, 'has_feedback', 1.0);
    if (label === 'up') {
      this.bump(targetQueryId, +1.0);
    }
    if (label === 'down') {
      this.bump(targetQueryId, -1.0);
    }
    if (label === 'retry') {
      this.bump(targetQueryId, -0.5);
    }
    this.dirty = true;
    this.evictIfNeeded();
    return id;
  }

  adjustWeights(nodeId: string, deltas: { feedbackDelta?: number; repeatDelta?: number }) {
    const n = this.nodes.get(nodeId);
    if (!n) {
      return;
    }
    if (deltas.feedbackDelta) {
      n.weights.feedback += deltas.feedbackDelta;
    }
    if (deltas.repeatDelta) {
      n.weights.repeat += deltas.repeatDelta;
    }
    n.lastAccess = Date.now();
    this.dirty = true;
  }

  link(from: string, to: string, type: EdgeType, weight: number) {
    const now = Date.now();
    if (!this.nodes.has(from) || !this.nodes.has(to)) {
      return;
    }
    const key = `${from}::${type}::${to}`;
    const prev = this.edges.get(key);
    const e: Edge = {
      id: key,
      from,
      to,
      type,
      weight: (prev?.weight ?? 0) + weight,
      createdAt: prev?.createdAt ?? now,
      lastAccess: now,
    };
    if (!prev) {
      const a = this.nodes.get(from)!;
      a.degree++;
      a.lastAccess = now;
      const b = this.nodes.get(to)!;
      b.degree++;
      b.lastAccess = now;
    }
    this.edges.set(key, e);
  }

  searchSimilarQueries(text: string, k = 5) {
    const emb = this.embedder.embed(text) as SparseVec;
    const out: { id: string; sim: number; score: number; text?: string; canonicalKey?: string }[] =
      [];
    const ids = new Set<string>();
    if (this.inverted) {
      const grams = normalize(text).split(' ');
      for (const g of grams) {
        for (const id of this.inverted.get(g) ?? []) {
          ids.add(id);
        }
      }
    } else {
      for (const id of this.nodes.keys()) {
        ids.add(id);
      }
    }
    for (const id of ids) {
      const n = this.nodes.get(id)!;
      if (n.type !== 'query' || !n.embedding) {
        continue;
      }
      const sim = cosineSparse(emb, n.embedding as SparseVec);
      const score = 0.7 * sim + 0.3 * this.rank(n);
      out.push({ id, sim, score, text: n.text, canonicalKey: n.canonicalKey });
    }
    return out.sort((a, b) => b.score - a.score).slice(0, k);
  }

  private bump(nodeId: string, delta: number) {
    const n = this.nodes.get(nodeId);
    if (!n) {
      return;
    }
    n.weights.feedback += delta;
    n.lastAccess = Date.now();
  }
  private touchNeighbors(nodeId: string, ts: number) {
    for (const e of this.edges.values()) {
      if (e.from === nodeId || e.to === nodeId) {
        e.lastAccess = ts;
      }
    }
  }
  private rank(n: BaseNode) {
    const now = Date.now();
    const ageMin = (now - n.lastAccess) / (1000 * 60);
    const recency = Math.exp(-Math.log(2) * (ageMin / this.cfg.halfLifeMin));
    const repeat = Math.tanh((n.weights.repeat ?? 0) / 5);
    const fb = Math.tanh((n.weights.feedback ?? 0) / 5);
    return 0.5 * recency + 0.3 * repeat + 0.2 * fb;
  }
  private evictIfNeeded() {
    const { maxNodes, maxEdges, protectMinDegree } = this.cfg;
    const overN = this.nodes.size - maxNodes;
    if (overN > 0) {
      const cand: BaseNode[] = [];
      for (const n of this.nodes.values()) {
        if (n.pinned) {
          continue;
        }
        if ((n.degree ?? 0) >= protectMinDegree) {
          continue;
        }
        cand.push(n);
      }
      cand.sort((a, b) => this.rank(a) - this.rank(b));
      for (let i = 0; i < overN && i < cand.length; i++) {
        this.removeNode(cand[i].id);
      }
    }
    const overE = this.edges.size - maxEdges;
    if (overE > 0) {
      const es = [...this.edges.values()].sort(
        (a, b) => a.weight - b.weight || a.lastAccess - b.lastAccess
      );
      for (let i = 0; i < overE && i < es.length; i++) {
        this.removeEdge(es[i].id);
      }
    }
  }
  private removeNode(id: string) {
    const n = this.nodes.get(id);
    if (!n) {
      return;
    }
    for (const e of [...this.edges.values()]) {
      if (e.from === id || e.to === id) {
        this.removeEdge(e.id);
      }
    }
    this.nodes.delete(id);
    if (n.canonicalKey) {
      this.byCanonical.delete(n.canonicalKey);
    }
  }
  private removeEdge(id: string) {
    const e = this.edges.get(id);
    if (!e) {
      return;
    }
    const a = this.nodes.get(e.from);
    if (a) {
      a.degree = Math.max(0, a.degree - 1);
    }
    const b = this.nodes.get(e.to);
    if (b) {
      b.degree = Math.max(0, b.degree - 1);
    }
    this.edges.delete(id);
  }

  stats() {
    return { nodes: this.nodes.size, edges: this.edges.size };
  }

  toSnapshot() {
    const nodes = [...this.nodes.values()].map((n) => ({
      ...n,
      embedding: serializeSparse(n.embedding),
    }));
    const edges = [...this.edges.values()];
    const embedder = this.embedder.exportState?.() ?? null;
    return { graph: { nodes, edges }, embedder, canonicalMeta: defaultCanonicalMeta };
  }
  fromSnapshot(s: { graph: { nodes: Array<{ id: string; embedding: unknown; [k: string]: unknown }>; edges: Edge[] }; embedder?: unknown }) {
    this.nodes.clear();
    this.edges.clear();
    this.byCanonical.clear();
    if (this.inverted) {
      this.inverted.clear();
    }
    for (const raw of s.graph.nodes) {
      const n = { ...(raw as Record<string, unknown>), embedding: deserializeSparse(raw.embedding) } as BaseNode;
      this.nodes.set(n.id, n);
      this.indexNode(n.id, n.text, n.canonicalKey);
    }
    for (const e of s.graph.edges) {
      this.edges.set(e.id, e);
    }
    this.embedder.importState?.(s.embedder);
  }
  async saveToFile(file: string, opts?: { onlyIfDirty?: boolean }) {
    if (opts?.onlyIfDirty && !this.dirty) {
      return;
    }
    await fs.writeFile(file, JSON.stringify(this.toSnapshot()), 'utf8');
    this.dirty = false;
  }
  static async loadFromFile(file: string, store: GraphStore) {
    const json = await fs.readFile(file, 'utf8');
    store.fromSnapshot(JSON.parse(json));
  }
}
