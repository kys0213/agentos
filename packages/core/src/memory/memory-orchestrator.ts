import path from 'path';
import fs from 'fs/promises';
import {
  GraphStore,
  SleepCycleBaseOptions,
  SleepCycleReport,
  SleepCycleRunOptions,
  SleepCycleSummarizer,
} from './graph-store';
import { GraphConfig, BaseNode, Edge } from './types';
import { SimpleEmbedding } from './embedding/simple-embedding';

export type Scope = 'agent' | 'session';

export interface OrchestratorCfg {
  sessionGraph: GraphConfig;
  agentGraph: GraphConfig;
  promotion: { minRank: number; maxPromotions: number; minDegree?: number; carryWeights?: boolean };
  checkpoint: { outDir: string; topK: number; pretty?: boolean };
  searchBiasSessionFirst?: number;
  sleepCycle?: {
    summarizer: SleepCycleSummarizer;
    defaults?: Partial<SleepCycleBaseOptions>;
    runOnFinalize?: boolean;
    runOptions?: SleepCycleRunOptions;
    emit?: (scope: Scope, report: SleepCycleReport) => void;
  };
}

export class MemoryOrchestrator {
  private embedder = new SimpleEmbedding();
  private agentStore: GraphStore;
  private sessions = new Map<string, GraphStore>();

  constructor(
    private agentId: string,
    private cfg: OrchestratorCfg
  ) {
    this.agentStore = new GraphStore({ ...cfg.agentGraph }, this.embedder);
    this.configureSleepCycle(this.agentStore, 'agent');
  }

  getSessionStore(sessionId: string) {
    if (!this.sessions.has(sessionId)) {
      const store = new GraphStore({ ...this.cfg.sessionGraph }, this.embedder);
      this.configureSleepCycle(store, 'session');
      this.sessions.set(sessionId, store);
    }
    return this.sessions.get(sessionId)!;
  }
  getAgentStore() {
    return this.agentStore;
  }
  getActiveSessions() {
    return [...this.sessions.keys()];
  }

  upsertQuery(sessionId: string, text: string) {
    return this.getSessionStore(sessionId).upsertQuery(text);
  }
  recordFeedback(sessionId: string, qId: string, label: 'up' | 'down' | 'retry', note?: string) {
    return this.getSessionStore(sessionId).recordFeedback(qId, label, note);
  }

  search(sessionId: string, text: string, k = 8) {
    const s = this.getSessionStore(sessionId);
    const sTop = s.searchSimilarQueries(text, k);
    const aTop = this.agentStore.searchSimilarQueries(text, k);
    const bias = this.cfg.searchBiasSessionFirst ?? 0.05;
    const map = new Map<
      string,
      { id: string; score: number; from: 'session' | 'agent'; text?: string; canonicalKey?: string }
    >();
    for (const r of sTop) {
      map.set(r.canonicalKey ?? r.id, {
        id: r.id,
        score: r.score + bias,
        from: 'session',
        text: r.text,
        canonicalKey: r.canonicalKey,
      });
    }
    for (const r of aTop) {
      const key = r.canonicalKey ?? r.id;
      const cur = map.get(key);
      if (!cur) {
        map.set(key, {
          id: r.id,
          score: r.score,
          from: 'agent',
          text: r.text,
          canonicalKey: r.canonicalKey,
        });
      } else {
        cur.score = Math.max(cur.score, r.score + bias);
      }
    }
    return [...map.values()].sort((a, b) => b.score - a.score).slice(0, k);
  }

  private configureSleepCycle(store: GraphStore, scope: Scope) {
    if (!this.cfg.sleepCycle) {
      return;
    }
    store.configureSleepCycle({
      summarizer: this.cfg.sleepCycle.summarizer,
      defaults: this.cfg.sleepCycle.defaults,
      emit: this.cfg.sleepCycle.emit
        ? (report) => this.cfg.sleepCycle?.emit?.(scope, report)
        : undefined,
    });
  }

  private rank(n: Pick<BaseNode, 'lastAccess' | 'weights'>, halfLifeMin: number) {
    const now = Date.now();
    const ageMin = (now - n.lastAccess) / (1000 * 60);
    const recency = Math.exp(-Math.log(2) * (ageMin / halfLifeMin));
    const repeat = Math.tanh((n.weights?.repeat ?? 0) / 5);
    const fb = Math.tanh((n.weights?.feedback ?? 0) / 5);
    return 0.5 * recency + 0.3 * repeat + 0.2 * fb;
  }
  private pickTopQueries(
    nodes: BaseNode[],
    minRank: number,
    topK: number,
    minDegree: number,
    halfLifeMin: number
  ) {
    return nodes
      .filter((n) => n.type === 'query')
      .map((n) => ({ n, r: this.rank(n, halfLifeMin) }))
      .filter((x) => x.r >= minRank && (x.n.degree ?? 0) >= minDegree)
      .sort((a, b) => b.r - a.r)
      .slice(0, topK)
      .map((x) => x.n);
  }
  private cpFile(name: string) {
    return path.join(this.cfg.checkpoint.outDir, name);
  }

  async finalizeSession(
    sessionId: string,
    opts?: { promote?: boolean; checkpoint?: boolean; checkpointName?: string }
  ) {
    const promote = opts?.promote ?? true;
    const checkpoint = opts?.checkpoint ?? true;
    const s = this.sessions.get(sessionId);
    if (!s) {
      return;
    }

    if (this.cfg.sleepCycle && this.cfg.sleepCycle.runOnFinalize !== false) {
      const overrides: SleepCycleRunOptions = {
        ...(this.cfg.sleepCycle.runOptions ?? {}),
        trigger: 'session-finalize',
        force: true,
      };
      s.runSleepCycle(overrides);
    }

    if (promote) {
      const promoted = this.promoteHotspotsFromSession(sessionId, {
        carryWeights: this.cfg.promotion.carryWeights,
      });
      if (promoted > 0) {
        await fs.mkdir(this.cfg.checkpoint.outDir, { recursive: true });
        await this.agentStore.saveToFile(this.cpFile(`agent-${this.agentId}.json`), {
          onlyIfDirty: true,
        });
      }
    }

    if (checkpoint) {
      const snap = s.toSnapshot();
      const top = this.pickTopQueries(
        this.normalizeNodes(
          snap.graph.nodes as Array<Omit<BaseNode, 'embedding'> & { embedding: unknown }>
        ),
        this.cfg.promotion.minRank,
        this.cfg.checkpoint.topK,
        this.cfg.promotion.minDegree ?? 0,
        this.cfg.sessionGraph.halfLifeMin
      );
      const summary = {
        version: 1,
        agentId: this.agentId,
        sessionId,
        createdAt: new Date().toISOString(),
        topQueries: top.map((q) => ({
          id: q.id,
          text: q.text,
          lastAccess: q.lastAccess,
          weights: q.weights,
          degree: q.degree,
          canonicalKey: q.canonicalKey,
        })),
        edges: (snap.graph.edges as Edge[])
          .filter((e) => top.some((q) => q.id === e.from || q.id === e.to))
          .map((e) => ({ from: e.from, to: e.to, type: e.type, weight: e.weight })),
        embedder: snap.embedder,
        canonicalMeta: snap.canonicalMeta,
      };
      await fs.mkdir(this.cfg.checkpoint.outDir, { recursive: true });
      const name =
        opts?.checkpointName ?? `checkpoint-${this.agentId}-${sessionId}-${Date.now()}.json`;
      await fs.writeFile(
        this.cpFile(name),
        JSON.stringify(summary, null, this.cfg.checkpoint.pretty ? 2 : 0),
        'utf8'
      );
    }
    this.sessions.delete(sessionId);
  }

  promoteHotspotsFromSession(sessionId: string, opts?: { carryWeights?: boolean }) {
    const s = this.sessions.get(sessionId);
    if (!s) {
      return 0;
    }
    const snap = s.toSnapshot();
    const top = this.pickTopQueries(
      this.normalizeNodes(
        snap.graph.nodes as Array<Omit<BaseNode, 'embedding'> & { embedding: unknown }>
      ),
      this.cfg.promotion.minRank,
      this.cfg.promotion.maxPromotions,
      this.cfg.promotion.minDegree ?? 0,
      this.cfg.sessionGraph.halfLifeMin
    );
    let count = 0;
    for (const n of top) {
      if (n.type !== 'query' || !n.text) {
        continue;
      }
      const id = this.agentStore.upsertQuery(n.text);
      if (opts?.carryWeights) {
        const delta = (n.weights?.feedback ?? 0) + (n.weights?.repeat ?? 0) * 0.2;
        this.agentStore.adjustWeights(id, { feedbackDelta: delta });
      }
      count++;
    }
    return count;
  }

  private normalizeNodes(
    nodes: Array<Omit<BaseNode, 'embedding'> & { embedding: unknown }>
  ): BaseNode[] {
    return nodes.map((n) => {
      const e = (n as { embedding?: unknown }).embedding;
      let embedding: number[] | Map<number, number> | undefined;
      if (e == null) {
        embedding = undefined;
      } else if (Array.isArray(e)) {
        if (e.length === 0 || typeof e[0] === 'number') {
          embedding = e as number[];
        } else {
          try {
            embedding = new Map(e as [number, number][]);
          } catch {
            embedding = undefined;
          }
        }
      } else if (e instanceof Map) {
        embedding = e as Map<number, number>;
      } else {
        embedding = undefined;
      }
      return { ...n, embedding } as BaseNode;
    });
  }
}
