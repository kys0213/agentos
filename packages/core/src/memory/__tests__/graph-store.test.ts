import { GraphStore } from '../../memory/graph-store';
import { SimpleEmbedding } from '../../memory/embedding/simple-embedding';

const cfg = { maxNodes: 1000, maxEdges: 4000, halfLifeMin: 240, tauDup: 0.96, tauSim: 0.75, protectMinDegree: 3, enableInvertedIndex: false };

describe('GraphStore', () => {
  test('upsertQuery deduplicates by canonicalKey (exact duplicate under normalization)', () => {
    const g = new GraphStore(cfg, new SimpleEmbedding());
    const a = g.upsertQuery('AgentOS MVP 설계');
    const b = g.upsertQuery('agentos mvp 설계'); // normalization -> same canonicalKey
    expect(b).toBe(a);
    expect(g.stats().nodes).toBe(1);
  });

  test('snapshot round-trip preserves searchable results', () => {
    const g1 = new GraphStore(cfg, new SimpleEmbedding());
    const q1 = g1.upsertQuery('에이전트 메모리 그래프 설계');
    g1.recordFeedback(q1, 'up', '좋아');
    const snap = g1.toSnapshot();

    const g2 = new GraphStore(cfg, new SimpleEmbedding());
    g2.fromSnapshot(snap);
    const before = g1.searchSimilarQueries('메모리 그래프 설계', 3);
    const after = g2.searchSimilarQueries('메모리 그래프 설계', 3);
    expect(after.length).toBeGreaterThan(0);
    // Compare top canonicalKey/text identity
    expect(after[0].canonicalKey).toBe(before[0].canonicalKey);
    expect(after[0].text).toBe(before[0].text);
  });
});

