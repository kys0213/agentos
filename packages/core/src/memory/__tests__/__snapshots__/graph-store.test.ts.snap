// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GraphStore > graph snapshot structure is stable (sanitized) [snapshot] 1`] = `
{
  "edgeTypeCounts": {
    "has_feedback": 3,
  },
  "nodes": [
    {
      "canonicalKey": "ck:nfc_v1:fnv1a64_demo:lu63c",
      "degree": 0,
      "text": "리그레이션 테스트 수행",
      "weights": {
        "feedback": 0,
        "repeat": 0,
      },
    },
    {
      "canonicalKey": "ck:nfc_v1:fnv1a64_demo:1ko3nin",
      "degree": 1,
      "text": "작성된 테스트케이스 조회",
      "weights": {
        "feedback": 1,
        "repeat": 0,
      },
    },
    {
      "canonicalKey": "ck:nfc_v1:fnv1a64_demo:8gtlil",
      "degree": 0,
      "text": "제품A 테스트케이스 검색",
      "weights": {
        "feedback": 0,
        "repeat": 0,
      },
    },
    {
      "canonicalKey": "ck:nfc_v1:fnv1a64_demo:1habwlu",
      "degree": 0,
      "text": "제품B 리팩토링 회귀 테스트 계획",
      "weights": {
        "feedback": 0,
        "repeat": 0,
      },
    },
    {
      "canonicalKey": "ck:nfc_v1:fnv1a64_demo:1ixda0d",
      "degree": 1,
      "text": "테스트 케이스 작성",
      "weights": {
        "feedback": 1,
        "repeat": 0,
      },
    },
    {
      "canonicalKey": "ck:nfc_v1:fnv1a64_demo:1ypkx2j",
      "degree": 1,
      "text": "테스트케이스 작성",
      "weights": {
        "feedback": 1,
        "repeat": 0,
      },
    },
  ],
}
`;
